test/matrix.lean

```diff
@@ -0,0 +1,169 @@
+/-
+manually ported from
+https://github.com/leanprover-community/mathlib/blob/4f4a1c875d0baa92ab5d92f3fb1bb258ad9f3e5b/test/matrix.lean
+-/
+import Mathlib.Data.Matrix.Notation
+-- import linear_algebra.Matrix.determinant
+import Mathlib.GroupTheory.Perm.Fin
+-- import Mathlib.Tactic.NormSwap
+import Qq
+
+open Qq
+
+-- todo: uncomment above imports when they are ported
+
+variables {α β : Type} [Semiring α] [Ring β]
+
+namespace Matrix
+
+open Matrix
+
+/-! Test that the dimensions are inferred correctly, even for empty matrices -/
+section dimensions
+
+-- set_option pp.universes true
+-- set_option pp.all true
+
+section elaborators
+open Lean Meta Elab Command
+
+/-- `dims% e` elaborates `e` as a Matrix and returns its dimensions as a `Nat × Nat`. -/
+elab "dims% " e:term : term => do
+  let elem_t ← mkFreshExprMVar (mkSort levelOne)
+  let m ← mkFreshExprMVar (mkConst ``Nat)
+  let n ← mkFreshExprMVar (mkConst ``Nat)
+  let matrix_t := mkAppN (mkConst ``Matrix [levelZero, levelZero, levelZero])
+                    #[mkApp (mkConst ``Fin) m, mkApp (mkConst ``Fin) n, elem_t]
+  let _ ← Term.elabTermEnsuringType e (some matrix_t)
+  let m ← instantiateMVars m
+  let n ← instantiateMVars n
+  mkAppM ``Prod.mk #[m, n]
+
+/-- `#guard e1 is e2` elaborates `e1` and `e2` and checks they are `BEq`. -/
+elab "#guard " e1:term " is " e2:term : command => liftTermElabM do
+  let ty ← mkFreshExprMVar none
+  let e1 ← Term.elabTerm e1 ty
+  let e2 ← Term.elabTermEnsuringType e2 ty
+  Elab.Term.synthesizeSyntheticMVarsUsingDefault
+  let e1 ← instantiateMVars e1
+  let e2 ← instantiateMVars e2
+  if e1.hasMVar then
+    throwError "expression{indentExpr e1}\nhas metavariables"
+  if e2.hasMVar then
+    throwError "expression{indentExpr e2}\nhas metavariables"
+  unless e1 == e2 do
+    throwError "expression{indentExpr e1}\nis not identically{indentExpr e2}"
+
+/-- `#guard e` elaborates `e` as a boolean and checks that it evaluates to `true`. -/
+elab "#guard " e:term : command => liftTermElabM do
+  let e ← Term.elabTermEnsuringType e (mkConst ``Bool)
+  Elab.Term.synthesizeSyntheticMVarsUsingDefault
+  let e ← instantiateMVars e
+  if e.hasMVar then
+    throwError "expression{indentExpr e}\nhas metavariables"
+  let v ← unsafe (evalExpr Bool (mkConst ``Bool) e)
+  unless v do
+    throwError "expression{indentExpr e}\ndid not evaluate to `true`"
+
+end elaborators
+
+-- we test equality of expressions here to ensure that we have `2` and not `1.succ` in the type
+#guard dims% !![]        is (0, 0)
+#guard dims% !![;]       is (1, 0)
+#guard dims% !![;;]      is (2, 0)
+#guard dims% !![,]       is (0, 1)
+#guard dims% !![,,]      is (0, 2)
+#guard dims% !![1]       is (1, 1)
+#guard dims% !![1,]      is (1, 1)
+#guard dims% !![1;]      is (1, 1)
+#guard dims% !![1,2;3,4] is (2, 2)
```

Comment:
@eric-wieser I just pushed a test framework that we can eventually generalize out of this file. The matrix-local part of it is this `dims%` elaborator.